persistence/sql/groups/persister_groups.go

```diff
@@ -0,0 +1,398 @@
+	OrganizationID uuid.UUID
+
+	// Filters is a list of SCIM filters to apply.
+	Filters []identity.SCIMFilter
+
+	// Count specifies the maximum number of items to return.
+	Count int
+
+	// StartIndex is the 1-based index of the first element in the list.
+	StartIndex int
+}
+
+type Persister interface {
+	CreateGroup(ctx context.Context, g *Group) error
+	ListGroups(ctx context.Context, params ListGroupsParameters) ([]Group, int, error)
+	GetGroup(ctx context.Context, id uuid.UUID) (*Group, error)
+	UpdateGroup(ctx context.Context, g *Group) ([]IdentityID, error)
+	DeleteGroup(ctx context.Context, id uuid.UUID) ([]IdentityID, error)
+	ListIdentityGroups(ctx context.Context, id uuid.UUID) ([]Group, error)
+}
+
+type dependencies interface {
+	contextx.Provider
+	x.TracingProvider
+}
+
+type GroupPersister struct {
+	dependencies
+	conn *pop.Connection
+	nid  uuid.UUID
+}
+
+type IdentityID struct {
+	ID uuid.UUID `db:"identity_id"`
+}
+
+func NewPersister(deps dependencies, conn *pop.Connection) *GroupPersister {
+	return &GroupPersister{
+		dependencies: deps,
+		conn:         conn,
+	}
+}
+
+func (p *GroupPersister) NetworkID(ctx context.Context) uuid.UUID {
+	return p.Contextualizer().Network(ctx, p.nid)
+}
+
+func (p *GroupPersister) WithNetworkID(nid uuid.UUID) *GroupPersister {
+	p.nid = nid
+	return p
+}
+
+func (p *GroupPersister) CreateGroup(ctx context.Context, g *Group) error {
+	g.NID = p.NetworkID(ctx)
+	return sqlcon.HandleError(popx.GetConnection(ctx, p.conn.WithContext(ctx)).Create(g))
+}
+
+var IndexedSCIMAttributes = map[string]string{
+	"displayName": "display_name",
+}
+
+func (p *GroupPersister) ListGroups(ctx context.Context, params ListGroupsParameters) (groups []Group, total int, err error) {
+	ctx, span := p.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.ListGroups",
+		trace.WithAttributes(attribute.Stringer("network.id", p.NetworkID(ctx))))
+	defer otelx.End(span, &err)
+
+	var (
+		query           strings.Builder
+		args            []any
+		nid             = p.NetworkID(ctx)
+		groupsWithTotal = make([]struct {
+			Group
+			Total int `db:"total"`
+		}, 0)
+	)
+
+	query.WriteString(`
+SELECT scim_groups.*, COUNT(*) OVER() AS total
+FROM scim_groups
+WHERE nid = ?
+`)
+	args = append(args, nid)
+
+	for _, scimFilter := range params.Filters {
+		// Do not allow filtering by SCIM attributes that are not indexed.
+		col, ok := IndexedSCIMAttributes[scimFilter.SCIMAttribute]
+		if !ok {
+			return nil, 0, errors.WithStack(herodot.ErrBadRequest.WithReasonf("The SCIM attribute %s is not indexed and cannot be used for filtering.", scimFilter.SCIMAttribute))
+		}
+
+		query.WriteString(fmt.Sprintf(`
+				AND %s = ?
+			`, col)) // This is not an SQL injection because we checked the SCIMAttribute above.
+		args = append(args, scimFilter.Value)
+	}
+
+	if !params.OrganizationID.IsNil() {
+		query.WriteString(`
+				AND organization_id = ?
+			`)
+		args = append(args, params.OrganizationID.String())
+	}
+	if params.Count > 0 {
```

Comment:
We must always enforce pagination, so the limit must be set to a fixed upper limit and never be unlimited.